### ISSUE TITLE
Fedora 34 Beta (GNOME 40) tray icon support package

### ISSUE THREAD

**Issue by RedBearAK** [rbreaves/kinto#472]
This is on Fedora 34 Beta (GNOME). Solution at the end. 

I was able to get the indicator tray extension working, now that some extensions have been updated for GNOME 40. Was able to install one application that displays a tray icon (Albert) so I knew the indicator tray was working. But the Kinto tray icon still wouldn't enable in the GUI or run from the terminal. The usual "namespace" problem. 

```
Traceback (most recent call last):
  File "/home/kris/.config/kinto/kintotray.py", line 6, in <module>
    gi.require_version('AppIndicator3', '0.1')
  File "/usr/lib64/python3.9/site-packages/gi/__init__.py", line 126, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace AppIndicator3 not available
```

Finally tracked the issue down to a missing package which is different from those listed by the installer currently for an apt-based system. There is no `gir1.2-appindicator3-0.1` package found on Fedora. 

But I found a reference online and this got the tray icon working: 

```
sudo dnf install libappindicator-gtk3
```

**Comment by rbreaves** [comment]
Oh that’s pretty awesome, thanks for tracking that down.. surely the standard Ubuntu gnome is the same.. need to get it updated too. I don’t like using the 3rd party extensions if it can be avoided.

**Comment by RedBearAK** [comment]
@rbreaves 

Looks like on Ubuntu 20.10 the `chrome-gnome-shell` package is required for the web browser extension to function. Fortunately, that gets pulled in as a dependency when installing probably `gnome-shell-extension-appindicator`. And contrary to Fedora, the `gir1.2-appindicator3-0.1` package is still present on Ubuntu 20.10. 

I'm going to also check the latest Ubuntu 21.04 daily to see if something has changed.

**Comment by RedBearAK** [comment]
@rbreaves 

Testing Ubuntu 21.04 (minimal install): 

Curl not installed by default. 

Using new install method. 

I don't know if this is going to be how it is when Ubuntu 21.04 releases, but once I ran the command at the end of Kinto install: 

```
sudo apt install gnome-tweaks gnome-shell-extension-appindicator gir1.2-appindicator3-0.1
```

I was then able to enable the tray icon in the Kinto GUI and it immediately appeared in the top bar. As in, _before_ I even tried to install the additional notifier extension from the web. 

They're still on GNOME 3.38 (and will be until next year, I think, or at least until the 21.10 release) but there is an application called Extensions installed by default [wrong, see below], and inside that is an extension called "Ubuntu AppIndicators", which was already enabled. So it "just worked" after installing the packages. 

So maybe Ubuntu decided people actually wanted their tray indicators back, and this won't be such a problem in the future. The `gir1.2-appindicator3-0.1` might actually be the only package necessary on 21.04 to get this working. 

Trying... and confirmed. If I do: 

```
sudo apt install curl gir1.2-appindicator3-0.1
```

... and then run the installer link, I can immediately enable the tray icon from the GUI app. But it kind of seems like it should be automatically enabled if the `gi` package is already there when I run the installer. Wonder why it isn't. 

OK, it looks like the Extensions app I saw before is actually not installed by default, it's provided by `gnome-shell-extension-prefs` which is pulled in by `gnome-tweaks`. Not necessary, but just making a note of it.

**Comment by rbreaves** [comment]
Well I know I saw a reviewer w/ my system tray running a little differently than I had under Ubuntu 20.04, the menu expanded differently & I’ve had no idea how they managed it.

**Comment by RedBearAK** [comment]
> Well I know I saw a reviewer w/ my system tray running a little differently than I had under Ubuntu 20.04, the menu expanded differently & I’ve had no idea how they managed it.

If you're talking about how the submenus expand vertically inside the menu instead of off to the side, that's happening already in 20.10. I don't have a 20.04 set up at the moment to see if it was different there. 

Here's hoping the need for the 3rd party extension will come to an end with 21.04. Just need to install that `gir1.2-appindicator3-0.1` package and auto-enable the tray icon.

**Comment by RedBearAK** [comment]
@rbreaves 

Ubuntu 20.10 also has the "Ubuntu AppIndicators" extension installed and active, so the only package that is actually needed to get the tray icon working is `gir1.2-appindicator3-0.1`. Literally nothing else. No 3rd party extension. I don't know if this is different from 20.04 and earlier. 

So it's just Fedora that needs the 3rd party extension.

**Comment by rbreaves** [comment]
My understanding is that they were working to actively remove appindicators for system tray apps in Gnome, but if they are reversing that then that would be a positive.

**Comment by RedBearAK** [comment]
@rbreaves 

My unfamiliarity with GNOME has led me slightly astray. Turns out the 3rd party extension is **_not_** necessary in Fedora. Technically. Although for most people that would still probably be the easiest way to activate the extension. Technically. 

Let me explain. Sorry for the wordage, but I want to be very clear on what I've figured out.  

I didn't realize the same `KStatusNotifierItem` I was installing from the website is the same extension installed by the gnome-shell-extension-appindicator package. For some reason I thought that package installed something different that didn't work, so I was sort of ignoring it. But I was messing with the `gnome-shell-extension-topicons-plus` package hoping to get the tray working with a minimum of fuss, when I noticed that `KStatusNotifierItem` was installed even though I had never gone to the extensions website this time around. So I finally realized where it comes from. No 3rd party extension necessary. Technically. 

BTW, I do not recommend the `gnome-shell-extension-topicons-plus` package. It will also be disabled by default on Fedora, and when you do enable it, by default it places the tray icons in the center of the top bar next to the clock, instead of in the standard location on the right. You have to open its preferences and change it. And the preferences dialog looks awful. At least it does right now, in GNOME 40. 

So why all the "technically"? Because, the `KStatusNotifierItem` extension that gets installed by the `gnome-shell-extension-appindicator` package is **_disabled_** by default, unlike the `Ubuntu AppIndicator` extension that comes pre-installed in Ubuntu, which is **_enabled_** by default. For a user that doesn't know how to activate GNOME extensions, it's basically useless. And apparently there is no easy way to see your extensions unless you know about the GNOME Extensions app, which is not pre-installed. 

So the user would need to A) know to look for and install the new GNOME Extensions app from the Software app, or B) know to install the `gnome-extensions-app` (in Fedora) package in the terminal. Then they would have to open the GNOME Extensions app and know to look for and manually enable the disabled `KStatusNotifierItem` extension. So simple! 

But, (this is the fun part) they can only do this **_after_** using Alt+F2 and entering the magical command "r" in the Run dialog to restart the GNOME shell, or after logging out and logging back in. Because the extension won't show up in the Extensions app until the shell is restarted. And won't disappear from the Extensions app if you uninstall the package. You have to quit the app and restart the shell and reopen the app. Every time. 

(What year is this, again?) 

Alternately, the user would have to know to do this from the command line to find and enable the extension, and of course know which extension they were even looking for: 

(`gnome-shell-extension-tool` is deprecated, replaced by `gnome-extensions`)

```
$ gnome-extensions list

apps-menu@gnome-shell-extensions.gcampax.github.com
background-logo@fedorahosted.org
launch-new-instance@gnome-shell-extensions.gcampax.github.com
places-menu@gnome-shell-extensions.gcampax.github.com
window-list@gnome-shell-extensions.gcampax.github.com
appindicatorsupport@rgcjonas.gmail.com

$ gnome-extensions enable appindicatorsupport@rgcjonas.gmail.com
```

It doesn't even show which extensions are enabled or disabled unless you get the detailed report with `list -d`. 

So there's a lot of specific knowledge needed to get the tray working with a minimum of effort. I wouldn't expect any newer user of GNOME to know how to do any of this. I sure didn't understand any of this until just now. If I hadn't been visiting some GNOME forum threads and seen someone talking about it, I also wouldn't know about the magical Alt+F2, "r" shortcut to restart the shell. 

You'll either have to do at least some of this from the installer or give slightly more elaborate instructions on how to get the tray working. Fortunately it's not too complicated from the terminal, once you know how. This should do it: 

(Fedora only): 
```
sudo apt install gnome-shell-extension-appindicator libappindicator-gtk3
busctl --user call org.gnome.Shell /org/gnome/Shell org.gnome.Shell Eval s 'Meta.restart("Restarting…")'
gnome-extensions enable appindicatorsupport@rgcjonas.gmail.com
```

And yes, I had to google that complicated `busctl` command just now. That is apparently exactly what happens in the background when you do the Alt+F2, "r" trick. So that should be the safest way to restart the shell from terminal. Seems to work fine, I just tested it in Fedora. 

I might run through the fresh install cycle one more time tomorrow just to make certain the `libappindicator-gtk3` package is really necessary, but I'm pretty sure that's still essential specifically to get `kintotray.py` to connect to the tray. That's what got the Kinto icon to appear before when I already had the Albert icon showing in the working tray, when I was using the `KStatusNotifierItem` extension installed from the web. So it should be necessary regardless of how the indicator tray itself gets activated. To get past the `AppIndicator3` "Namespace" error. 

tl;dr

Good news! No 3rd party extensions are needed on either Ubuntu 20.10 or later, or Fedora 34. Yay! 

In case it was not clear, current versions of the GNOME Tweaks app no longer deal with extensions, and it even brings up a dialog saying exactly what. Those are now completely isolated to the new GNOME Extensions app. This was true even in the Ubuntu 20.10 install, which is still using GNOME 3.38. The package name and icon was different, but it was the same Extensions app as the one on Fedora in GNOME 40.

**Comment by RedBearAK** [comment]
@rbreaves 

Confirming once again, `libappindicator-gtk3` is necessary to get `kintotray.py` to run and put the icon in the tray. The tray icon can't be enabled in the GUI app until that package is installed. It won't activate because of the AppIndicator3 "Namespace" error. 

Also confirmed, the `gnome-shell-extension-appindicator` package is NOT installed on Fedora by default, unlike Ubuntu. 

And these commands, in this order, is the minimum-effort method to get the tray working in a fresh Fedora GNOME install (copied from previous comment): 

```
sudo apt install gnome-shell-extension-appindicator libappindicator-gtk3
busctl --user call org.gnome.Shell /org/gnome/Shell org.gnome.Shell Eval s 'Meta.restart("Restarting…")'
gnome-extensions enable appindicatorsupport@rgcjonas.gmail.com
```

However at the moment (on Fedora) the tray icon still needs to be enabled manually in the Kinto GUI app, even if these packages are installed and the tray enabled prior to running the Kinto install. 

I guess I understand why you don't necessarily want to try and install these packages automatically, even with user input for permission, because some users will be using different extensions and there could be a conflict.

**Comment by sudhirkhanger** [comment]
This problem exists even on Kubuntu which can be remedied by installing `gir1.2-appindicator3-0.1`.
